File: web/area.js
/* Reading the area pack: the forecast for a point nobody listed in advance.

   The pipeline publishes a coarsened grid of temperature, precipitation and
   cloud cover, cut into tiles. Everything needed to find a point is in
   data/area/index.json, so nothing about the format is assumed here: the grid
   origin and step, the tile size, the order of the fields and their scales all
   come from the index. One tile is a few tens of kilobytes, so answering for a
   place costs a single small fetch. */
"use strict";

const AreaPack = (() => {
  const BASE = "data/area/";

  const READERS = {
    int16: (view, offset) => view.getInt16(offset, true),
    uint16: (view, offset) => view.getUint16(offset, true),
    uint8: (view, offset) => view.getUint8(offset),
  };
  const SIZES = { int16: 2, uint16: 2, uint8: 1 };

  let indexPromise = null;
  const tiles = new Map();

  /* The pack is not in the repository and reaches the site through the cache
     of the forecast workflow. It can therefore be missing, and the app has to
     say so rather than break. */
  class MissingPack extends Error {}
  class OffGrid extends Error {}

  function load(force) {
    const url = force ? `${BASE}index.json?t=${Date.now()}` : `${BASE}index.json`;
    return fetch(url, { cache: force ? "reload" : "no-store" }).then((response) => {
      if (!response.ok) throw new MissingPack(`HTTP ${response.status}`);
      return response.json();
    });
  }

  function index(force = false) {
    if (force || !indexPromise) {
      tiles.clear();
      indexPromise = load(force).catch((error) => {
        indexPromise = null;
        throw error;
      });
    }
    return indexPromise;
  }

  /* The nearest grid point, or null when the place lies outside the domain.
     Half a step of slack keeps a point on the very edge from being refused. */
  function nearest(pack, lat, lon) {
    const grid = pack.grid;
    const iy = Math.round((lat - grid.lat0) / grid.dlat);
    const ix = Math.round((lon - grid.lon0) / grid.dlon);
    if (iy < 0 || iy >= grid.ny || ix < 0 || ix >= grid.nx) return null;
    return {
      ix,
      iy,
      lat: grid.lat0 + iy * grid.dlat,
      lon: grid.lon0 + ix * grid.dlon,
    };
  }

  function distanceKm(fromLat, fromLon, toLat, toLon) {
    const R = 6371;
    const rad = Math.PI / 180;
    const dLat = (toLat - fromLat) * rad;
    const dLon = (toLon - fromLon) * rad;
    const a =
      Math.sin(dLat / 2) ** 2 +
      Math.cos(fromLat * rad) * Math.cos(toLat * rad) * Math.sin(dLon / 2) ** 2;
    return 2 * R * Math.asin(Math.sqrt(a));
  }

  async function tile(pack, tx, ty) {
    const name = `${tx}-${ty}.bin`;
    if (!tiles.has(name)) {
      const promise = fetch(BASE + name).then((response) => {
        if (!response.ok) throw new MissingPack(`HTTP ${response.status} for ${name}`);
        return response.arrayBuffer();
      });
      // A failed fetch must not be remembered as the answer for this tile.
      promise.catch(() => tiles.delete(name));
      tiles.set(name, promise);
    }
    return tiles.get(name);
  }

  /* One point out of a tile. Sections follow the order of the index and inside
     a section the hours of a point sit together, so a series is one walk with
     a fixed stride. */
  function decode(pack, buffer, width, height, row, column) {
    const view = new DataView(buffer);
    const hours = pack.hours;
    const point = row * width + column;
    const points = width * height;

    const values = {};
    let base = 0;
    for (const field of pack.fields) {
      const size = SIZES[field.type];
      const read = READERS[field.type];
      const series = new Array(hours);
      let offset = base + point * hours * size;
      for (let hour = 0; hour < hours; hour++, offset += size) {
        series[hour] = read(view, offset) / field.scale;
      }
      values[field.name] = series;
      base += points * hours * size;
    }
    return values;
  }

  /* The hours of one place, shaped like the rows of forecast.json so that the
     rest of the app cannot tell the two sources apart. Wind is not in the
     pack, so those fields are simply absent. */
  async function seriesAt(lat, lon) {
    const pack = await index();
    const grid = pack.grid;
    const at = nearest(pack, lat, lon);
    if (!at) throw new OffGrid("mimo doménu modelu");

    const tx = Math.floor(at.ix / pack.tile);
    const ty = Math.floor(at.iy / pack.tile);
    const width = Math.min(pack.tile, grid.nx - tx * pack.tile);
    const height = Math.min(pack.tile, grid.ny - ty * pack.tile);

    const buffer = await tile(pack, tx, ty);
    const values = decode(
      pack,
      buffer,
      width,
      height,
      at.iy - ty * pack.tile,
      at.ix - tx * pack.tile
    );

    const rows = pack.times.map((time, hour) => {
      const row = { time };
      for (const field of pack.fields) row[field.name] = values[field.name][hour];
      return row;
    });

    return {
      rows,
      point: at,
      distanceKm: distanceKm(lat, lon, at.lat, at.lon),
      runId: pack.run_id,
      generatedAt: pack.generated_at,
    };
  }

  async function covers(lat, lon) {
    return Boolean(nearest(await index(), lat, lon));
  }

  return { index, seriesAt, covers, MissingPack, OffGrid };
})();

File: pipeline/area.py
"""The area pack: a coarsened grid, cut into tiles, for points nobody listed.

The listed locations are extracted at the full resolution of the source and
published in data/forecast.json. That cannot answer for a place typed into the
app, so this module writes the other half: temperature, precipitation and cloud
cover over the whole domain, thinned to about two kilometres and split into
tiles the app fetches one at a time. A tile is a few tens of kilobytes, while
the whole pack is measured in megabytes, so the phone never pays for the
country to answer for one point.

Layout of a tile file. Sections follow the order of AREA_FIELDS with no padding
between them, and inside a section the whole time series of a point sits
together:

    index = (row * width + column) * hours + hour

Values are little-endian integers, and the value the app wants is the stored
integer divided by the scale of its field. Everything needed to read a tile -
grid origin and step, tile size, field order, scales, times - is in index.json
beside the tiles, so the app never has to guess the format.
"""

from __future__ import annotations

import json
import shutil
from datetime import datetime
from pathlib import Path

import numpy as np
import pandas as pd

from .config import AREA_FIELDS, AREA_TILE, AreaField

#: Storage types, spelled little-endian so the file does not depend on the
#: machine that wrote it.
DTYPES: dict[str, str] = {"int16": "<i2", "uint16": "<u2", "uint8": "|u1"}

INDEX_NAME = "index.json"


def align_grids(
    by_field: dict[str, tuple[np.ndarray, pd.DatetimeIndex]],
) -> tuple[dict[str, np.ndarray], pd.DatetimeIndex]:
    """Cut every field to the hours all of them share.

    Precipitation starts an hour later than the instantaneous fields, because
    an accumulation over a zero-length interval has no meaning. Without this
    the pack would carry an hour that only some of its fields cover.
    """
    if not by_field:
        raise ValueError("no fields to align")

    common: pd.DatetimeIndex | None = None
    for _, times in by_field.values():
        common = times if common is None else common.intersection(times)
    common = common.sort_values()
    if not len(common):
        raise ValueError("the fields share no valid time")

    aligned = {
        field: values[times.get_indexer(common)]
        for field, (values, times) in by_field.items()
    }
    return aligned, common


def encode_field(values: np.ndarray, spec: AreaField) -> np.ndarray:
    """Quantise one field to its storage type.

    Values are clipped to what the type holds rather than allowed to wrap
    around, so a stray reading can only be wrong by its own size, never by the
    width of the type.
    """
    dtype = np.dtype(DTYPES[spec.dtype])
    limits = np.iinfo(dtype)
    scaled = np.round(np.asarray(values, dtype="float64") * spec.scale)
    return np.clip(scaled, limits.min, limits.max).astype(dtype)


def tile_payload(
    encoded: list[tuple[AreaField, np.ndarray]], rows: slice, columns: slice
) -> bytes:
    """One tile: every field of it, each point's hours kept together."""
    chunks = []
    for _, array in encoded:
        window = array[:, rows, columns]
        # From [hour, row, column] to [row, column, hour], which in C order is
        # exactly the layout the app reads.
        chunks.append(np.moveaxis(window, 0, -1).ravel(order="C").tobytes())
    return b"".join(chunks)


def _format_time(moment: pd.Timestamp) -> str:
    return moment.tz_convert("UTC").strftime("%Y-%m-%dT%H:%MZ")


def build_index(
    run_id: str,
    generated_at: datetime,
    times: pd.DatetimeIndex,
    lats: np.ndarray,
    lons: np.ndarray,
    tile: int,
) -> dict:
    """The description the app reads before it touches a single tile."""
    return {
        "run_id": run_id,
        "generated_at": generated_at.strftime("%Y-%m-%dT%H:%M:%SZ"),
        "byte_order": "little",
        "grid": {
            "lat0": float(lats[0]),
            "lon0": float(lons[0]),
            "dlat": float(lats[1] - lats[0]),
            "dlon": float(lons[1] - lons[0]),
            "ny": int(len(lats)),
            "nx": int(len(lons)),
        },
        "tile": int(tile),
        "tiles": {
            "x": int(-(-len(lons) // tile)),
            "y": int(-(-len(lats) // tile)),
        },
        "hours": int(len(times)),
        "times": [_format_time(moment) for moment in times],
        "fields": [
            {"name": spec.field, "type": spec.dtype, "scale": spec.scale}
            for spec in AREA_FIELDS
        ],
    }


def write_area_pack(
    directory: Path,
    run_id: str,
    generated_at: datetime,
    grids: dict[str, np.ndarray],
    times: pd.DatetimeIndex,
    lats: np.ndarray,
    lons: np.ndarray,
    tile: int = AREA_TILE,
) -> dict:
    """Write index.json and one file per tile, and return the index.

    grids hold the fields in the units of data/forecast.json, indexed
    [hour, latitude, longitude]; precipitation is expected to be hourly
    already. The directory is replaced wholesale, so a pack never mixes tiles
    from two runs.
    """
    missing = [spec.field for spec in AREA_FIELDS if spec.field not in grids]
    if missing:
        raise ValueError(f"area pack is missing fields: {', '.join(missing)}")

    shape = (len(times), len(lats), len(lons))
    for field, values in grids.items():
        if values.shape != shape:
            raise ValueError(f"{field}: expected shape {shape}, got {values.shape}")

    encoded = [(spec, encode_field(grids[spec.field], spec)) for spec in AREA_FIELDS]
    index = build_index(run_id, generated_at, times, lats, lons, tile)

    if directory.exists():
        shutil.rmtree(directory)
    directory.mkdir(parents=True)

    for ty in range(index["tiles"]["y"]):
        rows = slice(ty * tile, min((ty + 1) * tile, len(lats)))
        for tx in range(index["tiles"]["x"]):
            columns = slice(tx * tile, min((tx + 1) * tile, len(lons)))
            payload = tile_payload(encoded, rows, columns)
            (directory / f"{tx}-{ty}.bin").write_bytes(payload)

    with (directory / INDEX_NAME).open("w", encoding="utf-8") as handle:
        json.dump(index, handle, ensure_ascii=False, indent=2)
        handle.write("\n")
    return index


def pack_size(directory: Path) -> int:
    """Bytes on disk, for the log line that says what a run published."""
    return sum(path.stat().st_size for path in directory.glob("*") if path.is_file())
